### ISSUE TITLE
`pysolve_ivp` can report simultaneous integration failure and success for large system sizes

### ISSUE THREAD

**Issue by dihm** [jrenaud90/CyRK#78]
Of course, by large, I mean very large. Below is the code to reproduce

```python
import numpy as np
import numba as nb
from CyRK import pysolve_ivp
import math

# define base system
y_0 = 1
t_span = (0.0, 8.0)
@nb.njit("void(f8[::1], f8, f8[::1]")
def cy_diffeq(dy, t, y):
    for i in range(y.size):
        dy[i] = -y[i]/2 + 2*math.sin(3*t)
```
This code works as expected
```python
# simultaneously solve many copies of this system
Neqs = 2**21
print(Neqs)
y0 = np.full(Neqs, y_0, dtype=np.float64)
result_max = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45", rtol=1e-7, atol=1e-7, pass_dy_as_arg=True)
print(result_max.success, result.message)
```
```shell
2097152
True Integration completed without issue
```

This block gives conflicting reports on integration success, but result appears to be correct
```python
Neqs = 2**22
print(Neqs)
y0 = np.full(Neqs, y_0, dtype=np.float64)
result_max = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45", rtol=1e-7, atol=1e-7, pass_dy_as_arg=True)
print(result_max.success, result.message)
```

```shell
4194304
False Integration completed without issue
```

Dependencies (via conda)
```shell
cyrk==0.11.4
python==3.11.5
numba==0.58.1
numpy==1.26.4
```

**Comment by jrenaud90** [comment]
Hey @dihm - Thanks for the report. 

So I was able to get this to run okay locally. I noticed an issue in the version posted above. The result message printed for the 2nd example is actually for the first integration (look at the usage of `result_max` vs. `result`). The 2nd one still fails for me but using the right result message we can see why. Here are my edits with the outputs.

```python
# simultaneously solve many copies of this system
Neqs = 2**21
print(Neqs)
y0 = np.full(Neqs, y_0, dtype=np.float64)
result = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45", rtol=1e-7, atol=1e-7, pass_dy_as_arg=True)
print(result.success, result.message)
print("Steps Required:", result.steps_taken)
```
```
2097152
True Integration completed without issue.
Steps Required: 61
```
and
```python
Neqs = 2**22
print(Neqs)
y0 = np.full(Neqs, y_0, dtype=np.float64)
result_larger = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45", rtol=1e-7, atol=1e-7, pass_dy_as_arg=True)
print(result_larger.success, result_larger.message)
print("Steps Required:", result_larger.steps_taken)
```
```
4194304
False Maximum number of steps (set by system architecture) exceeded during integration.
Steps Required: 58
```
We can see the last one failed because we exceeded the limit imposed by the optional argument `max_ram_MB`. This exisits to auto fail integration if the solver is taking up too much memory. It is more for folks using CyRK in MCMC schemes where they are running 10,000s of runs at once and would rather have 1 solver fail than have it eat 99% of the server memory. In this case once it gets to step 59, it is storing 59 (steps) * 4194304 (num y) * 8 (bytes for doubles) = 1979 MB which, with some other overhead in other arrays, brings us over the default limit of 2000MB. But if your system can take it, it is easy to adjust this limit. The following code completed successfully for me. 
```python
Neqs = 2**22
print(Neqs)
y0 = np.full(Neqs, y_0, dtype=np.float64)
result_more_steps = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45", rtol=1e-7, atol=1e-7, pass_dy_as_arg=True, max_ram_MB=8_000)
print(result_more_steps.success, result_more_steps.message)
print("Steps Required:", result_more_steps.steps_taken)
```
```
4194304
True Integration completed without issue.
Steps Required: 61
```

**Comment by jrenaud90** [comment]
This does not address the issue discussed in the other thread (#74 ) where you were finding seg faults when using large number of diffeqs. I am copying over that convo here so we can use this to track that problem. I will also add some tests in CyRK v0.12.0 (releasing later today) to look at large N.

Previous convo:

> @dihm: Hmm. My simple plan to solve a bunch of simple diffeqs in parallel has not triggered the issue (though I may have found a case where the result success reports False but the message and actual output appear to have worked correctly, with Neqs>2**21).
> 
> I am now more strongly suspicious of the code on my end being the likely culprit. I'll have to do some more debugging to see if I can track down what the issue is.

> @jrenaud90: Oh no! That is odd that it is happening with such a specific number. If you could make an example that replicates the crash then I can take a look at it. And I will integrate it into cyrks test suite to catch these in the future.
> My guess is something is not being resized correctly. Or, perhaps worse to debug, it is trying to resize but the memory is not available. I don't think that is likely though.

> @dihm: Hmm. My simple plan to solve a bunch of simple diffeqs in parallel has not triggered the issue (though I may have found a case where the result success reports False but the message and actual output appear to have worked correctly, with Neqs>2**21).
> I am now more strongly suspicious of the code on my end being the likely culprit. I'll have to do some more debugging to see if I can track down what the issue is.

**Comment by dihm** [comment]
Thanks Joe for fixing my dumb mistakes. The memory size issue totally makes sense and is obviously not a real issue.

I'm still working on reproducing my original problem. I've made some headway, but still can't quite reproduce outside my specific use case yet. Whenever I get a better example, I'll re-title to be a bit more helpful. 

The short of what I have at this exact moment is that pysolve_ivp is working correctly, but when I pass out a view to the reshaped `result.y` array via a function return, the memory is _somehow_ getting corrupted. Any access to the array memory causes a silent crash of the interpreter. Again, I'm having trouble isolating the root cause and even some of the triggers are finicky (ie I have more  crashes in ipynb than with the standard shell). But it does appear that creating a copy of the `result.y` array before passing out of the owning function namespace can fix the issue.

**Comment by jrenaud90** [comment]
Ah! I think you have hit the problem on the head @dihm. The data stored after integration is stored in arrays held by the result instance. If that instance dies then the memory of the arrays dies with it. 

So if you do:
```python
def f():
    result = pysolve_ivp(...)
    y = result.y
    return y
y_arr = f()
print(y_arr)
```
The data that y_arr was pointing to is no longer there. Best case scenario it is still held by the program and it may even still have some of its old values - but unexpected results are... very much expected. Or worst case that memory has been freed and you get a crash because you are trying to access memory that python no longer controls. This inconsistency may be why it is hard to reproduce.

The workaround as you said is to make a copy of the array and return it. 
```python
def f():
    result = pysolve_ivp(...)
    y = np.copy(result.y)
    return y
y_arr = f()
print(y_arr)
```
Or if you want a small performance boost you can avoid the copy and just ensure that result stays alive.
```python
def f():
    result = pysolve_ivp(...)
    return result
my_result = f()
print(my_result.y)
```

The copying fix could be done by cyrk automatically (at that performance hit). I think the performance hit is better than a crash though. So, I will open a new issue specifying this problem and can add a fix for the next version. 

I will leave this issue open for now just to make sure that is the real problem. If you can do some testing where you return y and return y's copy and see if you get crashes with the first and none with the second then I think we can confirm that is the problem.

**Comment by dihm** [comment]
Should have kept going a bit further. Here is a test that results reliably in the error. Exactly which iteration is fails on does seem to be system and interpreter dependent. Run as a script via CPython I'm getting up to 34. Run with a jupyter kernel I get failure around i=70. Forcing a copy of `result.y` prevents the error.

```python
import numpy as np
import numba as nb
from CyRK import pysolve_ivp
import math

y_0 = 1

t_span = (0.0, 8.0)
t_num = 100
t_eval = np.linspace(*t_span, num=t_num, dtype=np.float64)

@nb.njit("void(f8[::1], f8, f8[::1])")
def cy_diffeq(dy, t, y):

    for i in range(y.size):
        dy[i] = -y[i]/2 + 2*math.sin(3*t)

def solve_system(s=1, b=1):

    Neqs = b*s

    y0 = np.full(Neqs, y_0, dtype=np.float64)

    result = pysolve_ivp(cy_diffeq, t_span, y0, method="RK45",
                        rtol=1e-7,
                        atol=1e-7,
                        t_eval=t_eval,
                        pass_dy_as_arg=True)
    
    return result.y.reshape((s,b,t_num))

# increment system size until things crash
for i in range (2,100):
    print(f'Solving s={i}', end='\n')
    result_t = solve_system(s=i, b=15)
    count = np.count_nonzero(result_t) # this is the line that crashes
```

**Comment by jrenaud90** [comment]
Fixed with PR #81 in v0.12.1
